docs: decision record about key management (#4736)
eclipse-edc/Connector: docs/developer/decision-records/2025-01-17-key-management-improvement/README.md

is typically implemented by HSMs such as Hashicorp Vault.

In a similar fashion EDC will provide a feature to "remotely sign" payloads. Instead of transmitting the private key
from the vault to EDC over the network, and signing the payload locally, which carries some risk of exposure, leakage,
mishandling etc.,
EDC will transmit the payload to the vault, have the payload signed/encrypted there and transmit the signature or
encrypted payload back.

With this, the private key never leaves the vault, thus avoiding any risk of key exposure or leakage.

> _Note that there may be a penalty on latency, especially when transmitting very large payloads, such as
VerifiablePresentations that contain many large VerifiableCredentials!_

## Approach

### Implementation Remote signing

- add new interface `SigningService` that declares two methods:
  ```java
  public interface SigningService {
    byte[] sign(String keyId, byte[] data);  
    boolean verify(String keyId, byte[] signature);
  }
  ```
- the default implementation performs signing and verifying exactly as implemented currently, pulling keys out of the

>   vault and performing locally.

Reviewer: ```suggestion
  vault and performing the signing locally.
```

> - for each vault technology, there would be an implementation

Reviewer: ```suggestion
- For each vault technology, there would be a `SigningService` implementation
```

> - implement custom `JWSSigner` and `JWSVerifier` (from Nimbus) named `RemoteJwsSigner` and `RemoteJwsVerifier`

Reviewer: ```suggestion
- Custom implementations of `JWSSigner` and `JWSVerifier` (from Nimbus) named `RemoteJwsSigner` and `RemoteJwsVerifier`
```

>   respectively, that delegate signing and verifying to the `SigningService`.

Reviewer: ```suggestion
  respectively, will delegate signing and verifying to the `SigningService`.
```

  These objects are provided by an extension.

> - configuration values: `edc.signing.remote.enable` to enable/disable remote signing

Reviewer: ```suggestion
- A configuration value `edc.signing.remote.enable` will be added to enable/disable remote signing
```

> - A configuration value `edc.signing.remote.enable` will be added to enable/disable remote signing

Reviewer 2: Small nit:
Isn't this configuration redundant? If you add a remote signing service to your runtime, you'd expect remote signing to be done.
I could imagine a case where the remote signing service is added, only for the edc operator to find out later it wasn't working because it wasn't activated...

### Implementation Key Rotation

> - add `rotate(String keyId, Map<String, Object> keyProperties)` method to the `SigningService` to specify the algorithm

Reviewer: ```suggestion
- Add `rotate(String keyId, Map<String, Object> keyProperties)` method to the `SigningService` to specify the algorithm
```

  and cryptographic properties as well as the TTL (time-to-live)
- IdentityHub: `KeyPairServiceImpl` calls `SigningService#rotate()` when rotating keys

> - configuration values:

Reviewer: ```suggestion
- Configuration values:
```

- remote signing: it should be possible to cryptographically sign payloads "remotely", i.e. outside of EDC

## Rationale

Production-grade enterprise deployments of the Connector and other components will require 
centralized key management capabilities, where key pairs are automatically rotated to defend against key attrition. This
is typically implemented by HSMs such as Hashicorp Vault.

In a similar fashion EDC will provide a feature to "remotely sign" payloads. Instead of transmitting the private key
from the vault to EDC over the network, and signing the payload locally, which carries some risk of exposure, leakage, mishandling etc.,
EDC will transmit the payload to the vault, have the payload signed/encrypted there and transmit the signature or encrypted payload back.

With this, the private key never leaves the vault, thus avoiding any risk of key exposure or leakage.

> _Note that there may be a penalty on latency, especially when transmitting very large payloads, such as
VerifiablePresentations that contain many large VerifiableCredentials!_

## Approach

### Implementation Remote signing

- add `byte[] sign(String keyId, byte[] data)` method to `Vault`

> - add `boolean verify(String keyId, byte[] signature)` method to `Vault`

Author: Coupling the signing process to the Vault isn't necessarily a bad thing, because it reflects reality: most Vaults (Azure, AWS, HCV) support facilities to sign, verify and rotate, in addition to be secure Key-Value stores.

So to me that coupling, and introducing a `DigitalSignatureService` is not mutually exclusive: the `DigitalSignatureService` would simply be the "client code" that calls methods on the `Vault`. To do that, the vault would still have to expose `sign`, `verify` and `rotate` methods. 

This would mean, there could be two `DigitalSignatureServices`: 
- a default one, that pulls secrets and signs/verifies locally
- a `VaultSignatureService`, that delegates those calls to the `Vault`

> - add `boolean verify(String keyId, byte[] signature)` method to `Vault`

Reviewer 3: > [edit]: it would be the `DigitalSignatureService` implementation that uses `JWSSigner` and `JWSVerifier` objects from Nimbus. IF we went with separate services _per Vault_, this would introduce a dependency onto Nimbus in each Vault.

it's the other way around: the "nimbus way" to sign will just need to be implemented once in a "DigitalSignatureService" implementation (registered by default and eventually overridable), instead to need to be provided for every `Vault` implementation (because the `sign`, `verify` and `rotate` will need separate implementations for different `Vault` implementation, but, at the end, they will be the same).

> - add `boolean verify(String keyId, byte[] signature)` method to `Vault`

Author: Nimbus is mainly relevant for JWTs, and for that, it is sufficient to provide custom JWSSigner and JWSVerifier objects that we can plug in. We already have a `TokenGenerationService`, that can be customized with a `JWSSigner`, all we need is to do the same for the `TokenValidationService`.

In cases, where non-JWT content is signed, we could use such a DigitalSignatureService`. I don't think we have such a case ATM though.

> - add `boolean verify(String keyId, byte[] signature)` method to `Vault`

Reviewer 3: I was more referring to changing the `Vault` interface, that doesn't look the best approach, because signature (also if only JWT) is something that "relies" on vault to sign or it can be done by a vault service, but also by a separate service.

An example is how AWS manages this: Secret Manager keeps secrets, while the signature as a service is made by a different product called KMS.

Keeping the interfaces separated will help flexibility and avoid hard coupling between vault and a signature service

> - add `boolean verify(String keyId, byte[] signature)` method to `Vault`

Author: To sum up: 

If we keep this separate, there would have to be a `AwsKmsSignatureService`, a `AzureVaultSignatureService`, a `HashicorpVaultSignatureService`, a `GcpSecretsManagerSignatureService`, etc.

Combining this, the `JWSSigner` and `JWSVerifier` would delegate to the `SignatureService` (instead of the Vault).

The default implementation of the SignatureService would simply take a vault, pull the keys out and sign/verify locally.

do i have this correct so far?

> - add `boolean verify(String keyId, byte[] signature)` method to `Vault`

Reviewer 3: Yes, looks good :+1:

> - add `boolean verify(String keyId, byte[] signature)` method to `Vault`

Author: wanna give it another look?
